app/controllers/doorkeeper/tokens_controller.rb

```diff
@@ -43,31 +50,35 @@ def revoke
     # types, they set the application_id as null (since the claim cannot be
     # verified).
     #
-    # https://tools.ietf.org/html/rfc6749#section-2.1
-    # https://tools.ietf.org/html/rfc7009
+    # https://datatracker.ietf.org/doc/html/rfc6749#section-2.1
+    # https://datatracker.ietf.org/doc/html/rfc7009
     def authorized?
-      if token.present?
-        # Client is confidential, therefore client authentication & authorization
-        # is required
-        if token.application_id?
-          # We authorize client by checking token's application
-          server.client && server.client.application == token.application
-        else
-          # Client is public, authentication unnecessary
-          true
-        end
+      # Token belongs to specific client, so we need to check if
+      # authenticated client could access it.
+      if token.application_id? && token.application.confidential?
```

Comment:
If _either_ of `token.application_id?` or `token.application.confidential?` is not truthy, this method returns `true` (ie. "authorized == true")
Is that the intended behavior?